domain + application логики пользователя

tanyginvv opened the pull request.

tanyginvv committed:
добавил domain + application логики пользователя с правки по преддзащите

VladDotMe commented on backend/Recipes/Recipes.Domain/Entities/Recipe.cs:
```diff
@@ -2,17 +2,21 @@
 public class Recipe : Entity
 {
+    public int UserId { get; set; }
+    public User User { get; set; }
```
Author и AuthorId лучше, User непонятно

VladDotMe commented:
Жду полную версию, идея в том, чтобы проект всегда компилировался и работал заявленный функционал. 

Если делить пуллреквесты - то только по фичам(например, фича авторизации и все что с ней связано полностью), а не по слоям.

tanyginvv committed:
доделал полную версию

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,136 @@
+[Route( "api/users" )]
+public class UsersController() : ControllerBase
+{
+
```
тут проблемы с форматированием и () лишнее

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/AuthorizationController.cs:
```diff
@@ -0,0 +1,74 @@
+
+[ApiController]
+[Route( "api/users" )]
+public class AuthenticationController() : ControllerBase
```
В названии Authentication, а здесь Authorization. 
Знаешь чем отличается аутентификация от авторизации?

А контроллер лучше все-таки назвать UsersController.

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/AuthorizationController.cs:
```diff
@@ -0,0 +1,74 @@
+{
+    [HttpPost]
+    [Route( "registrate" )]
+    public async Task<IActionResult> Registrate( [FromBody] RegistrateUserDto registrateUserDto,
```
register

VladDotMe commented on backend/Recipes/Recipes.WebApi/Dto/AuthenticationDto/AuthenticationDto.cs:
```diff
@@ -0,0 +1,7 @@
+public class AuthenticationDto
+{
+    public string Login { get; init; }
+    public string PasswordHash { get; init; }
```
У тебя не соответствует тут название и логика в сервисе. Ты передаешь PasswordHash но в сервисе сравниваешь через метод BCrypt оригинальный пароль и хешированную версию из User, хотя по факту у тебя уже хешированный приходит. Мы с Димой обсудили потом, лучше с фронтенда передавать простой пароль, не хешированный, убери Hash из названия просто, по остальной логике все ок.

VladDotMe commented on backend/Recipes/Recipes.WebApi/Dto/UserDtos/RegistrateUserDto.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public string Name { get; init; }
+    public string Login { get; init; }
+    public string Description { get; init; }
```
при регистрации ты не вводишь описание, поэтому поле ненужно.

VladDotMe commented on backend/Recipes/Recipes.WebApi/Dto/UserDtos/RegistrateUserDto.cs:
```diff
@@ -0,0 +1,10 @@
+﻿
+namespace Recipes.WebApi.Dto.UserDto;
```
namespace некорректный и в других местах тоже.

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Users/Commands/CreateUser/CreateUserCommandHandler.cs:
```diff
@@ -0,0 +1,34 @@
+using Recipes.Application.Validation;
+using Recipes.Domain.Entities;
+
+namespace Recipes.Application.UseCases.Users.Commands;
```
namespace

VladDotMe commented on backend/Recipes/Recipes.WebApi/Dto/UserDtos/RegistrateUserDto.cs:
```diff
@@ -0,0 +1,10 @@
+
+public class RegistrateUserDto
+{
+    public string Name { get; init; }
```
Здесь бы тоже валидацию как в других DTO делал, Required там аттрибуты и на длину

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -18,6 +19,7 @@ namespace Recipes.WebApi.Controllers;
 {
+    [JwtAuthorization]
     [HttpPost]
     public async Task<ActionResult<RecipeReadIdDto>> CreateRecipe(
```
видимо не заметил сразу, imageTool здесь точно не должен использоваться, зачем здесь удалять изображение?

VladDotMe commented on backend/Recipes/Recipes.Domain/Entities/User.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public string Name { get; set; }
+    public string Description { get; set; }
+    public string Login { get; set; }
```
все сеттеры private, для ефки этого хватит. Возможно даже get; init; хватит, но не уверен, проверь)

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/AuthorizationController.cs:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    [HttpPost( "authentication" )]
+    public async Task<IActionResult> Authentication( [FromBody] AuthenticationDto authenticationDto,
```
login будет понятнее, думаю)

Еще укажи конкретный возвращаемый тип а не IActionResult, в соседних местах аналогично.

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Services/AuthService/AuthenticationService.cs:
```diff
@@ -0,0 +1,59 @@
+            return Result<AuthenticateUserDto>.FromError( "Введен неверный пароль" );
+        }
+
+        UserAuthorizationToken token = await userAuthorizationTokenRepository.GetByUserIdAsync( user.Id );
```
А почему AuthorizationToken а не Authentication?) Предлагю просто назвать UserAuthToken чтобы не возникало таких вопросов)

VladDotMe commented on backend/Recipes/Recipes.Domain/Entities/UserAuthorizationToken.cs:
```diff
@@ -0,0 +1,16 @@
+﻿namespace Recipes.Domain.Entities;
+
+public class UserAuthorizationToken
```
UserAuthToken

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/ConfigurationUtils/TokenConfiguration.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Recipes.Infrastructure.ConfigurationUtils;
+
+public class TokenConfiguration( IConfiguration configuration ) : ITokenConfiguration
```
Давай использовать паттерн IOptions

VladDotMe replied:
После добавления IOptions, класс тоже лишним получается..

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Services/AuthService/AuthenticationService.cs:
```diff
@@ -0,0 +1,59 @@
+        string accessToken = tokenCreator.GenerateAccessToken( user.Id );
+        string refreshToken = tokenCreator.GenerateRefreshToken();
+
+        DateTime refreshTokenExpiryDate = DateTime.Now.AddDays( tokenConfiguration.GetRefreshTokenValidityInDays() );
```
UtcNow

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/UserAuthorizationTokens/RefreshToken/RefreshTokenCommandHandler.cs:
```diff
@@ -0,0 +1,54 @@
+        UserAuthorizationToken token = await userAuthorizationTokenRepository.GetByRefreshTokenAsync( command.RefreshToken );
+        await userAuthorizationTokenRepository.Delete( token );
+
+        string accessToken = tokenCreator.GenerateAccessToken( token.UserId );
```
смотри, у тебя этот метод практчиески 1в1 копирует логику из метода AuthenticateUserAsync, различные только входные параметры и способ получения токена. Предлагаю сделать так, чтобы тот сервис принимал на вход UserAuthorizationToken и возвращал TokenDto. И это не команда а тоже должен быть сервис.


Итого, должно получиться:
1. RefreshTokenService - валидирует вход, получает UserAuthorizationToken, вызывает AuthTokenService с userId
2. LoginService , принимает login, password, валидирует, вызывает AuthTokenService  c userId
3. AuthTokenService -> принимает на вход userId  и включает в себя всю логику (подчистка токенов, создание нового токена и т.д.)

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,136 @@
+
+    [JwtAuthorization]
+    [HttpDelete( "{id}" )]
+    public async Task<IActionResult> DeleteUser(
```
это для кого метод?) в дизайне вроде такого функционала не было?

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,136 @@
+
+    [JwtAuthorization]
+    [HttpGet( "{userId}/recipes" )]
+    public async Task<IActionResult> GetRecipes(
```
это должно быть в RecipesController, а UserId прокидыватся в спецификацию метода GetRecipes

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,136 @@
+    [JwtAuthorization]
+    [HttpPut( "{id}" )]
+    public async Task<IActionResult> UpdateUser(
+        int id,
```
напиши явно аттрибуты, [FromRoute], [FromBody]

VladDotMe commented on backend/Recipes/Recipes.WebApi/Dto/UserDtos/UpdateUserDto.cs:
```diff
@@ -0,0 +1,10 @@
+    public string Name { get; init; }
+    public string Description { get; init; }
+    public string Login { get; init; }
+    public string OldPasswordHash { get; init; }
```
не хеш а обычный пароль, у тебя хеша на фронтенде нет и не должно быть

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Users/Commands/UpdateUser/UpdateUserCommandHandler.cs:
```diff
@@ -0,0 +1,50 @@
+            user.Login = command.Login;
+        }
+
+        if ( !string.IsNullOrEmpty( command.OldPasswordHash ) && !string.IsNullOrEmpty( command.NewPasswordHash ) )
```
а где проверка, что старый пароль - верный?)

tanyginvv replied:
В валидаторе

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,136 @@
+
+    [JwtAuthorization]
+    [HttpPut( "{id}" )]
+    public async Task<IActionResult> UpdateUser(
```
Тоже возвращай типы конкретные везде в Task<..>

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    [HttpGet( "{id}" )]
+    public async Task<ActionResult<UserDto>> GetUserById( int id,
```
GetUser, без id, просто отдавать текущего юзера

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,136 @@
+
+[ApiController]
+[Route( "api/users" )]
+public class UsersController() : ControllerBase
```
объединить контроллер с другим

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,136 @@
+{
+
+    [HttpGet( "login/{id}" )]
+    public async Task<ActionResult<ReadUserDto>> GetUserLoginById( int id,
```
удалить метод, вместе с рецептом отдавать логин автора

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommand.cs:
```diff
@@ -4,6 +4,7 @@ namespace Recipes.Application.UseCases.Recipes.Commands.CreateRecipe;
 
 public class CreateRecipeCommand
 {
+    public required int UserId { get; init; }
```
что-то не очень секьюрно) это получается можно любой юзерайди прокинуть и ты создашь от другого чувака) нужно отсюда убрать UserId и брать его из JWT Token'a

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -36,12 +38,18 @@ public async Task<ActionResult<RecipeReadIdDto>> CreateRecipe(
     [HttpDelete( "{id}" )]
     public async Task<IActionResult> DeleteRecipe(
         [FromRoute, Range( 1, int.MaxValue )] int id,
+        [FromQuery] int userId,
```
аналогично, нельзя прокидывать userId, надо доставать из Claims

VladDotMe commented on backend/Recipes/Recipes.WebApi/Dto/RecipeDtos/RecipeUpdateDto.cs:
```diff
@@ -7,6 +7,9 @@ namespace Recipes.WebApi.Dto.RecipeDtos;
 public class RecipeUpdateDto
 {
+    [Required]
+    public int UserId { get; set; }
```
убрать

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -92,13 +101,15 @@ public async Task<ActionResult<RecipeReadDto>> GetRecipeById(
     [HttpGet]
     public async Task<ActionResult<IEnumerable<RecipePartReadDto>>> GetRecipes(
         [FromServices] IQueryHandler<IEnumerable<GetRecipePartDto>, GetRecipesQuery> getRecipesQueryHandler,
+        [FromRoute] int userId = 0,
```
тут тоже можно убрать userId и доставать из claims

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandValidator.cs:
```diff
@@ -43,6 +43,11 @@ public async Task<Result> ValidateAsync( CreateRecipeCommand command )
 
+        if ( command.Tags.Count > 5 )
+        {
+            return Result.FromError( "Количество тегов ограничено до 5 " );
```
на фронтенде не забудь тоже ограничить)

tanyginvv committed:
вынес логику логинации и обновления токена в сервисы + остальные правки

tanyginvv committed:
поправил сеттеры

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/ConfigurationUtils/TokenConfiguration.cs:
```diff
@@ -1,25 +1,22 @@
 namespace Recipes.Infrastructure.ConfigurationUtils;
 
-public class TokenConfiguration( IConfiguration configuration ) : ITokenConfiguration
+public class TokenConfiguration(IOptions<JwtOptions> jwtOptions )  : ITokenConfiguration
```
пробел

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -12,26 +10,63 @@
+        {
+            return Unauthorized();
+        }
+        CreateRecipeCommand command = new()
```
это через мапстер создавать

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -40,15 +75,20 @@ public async Task<ActionResult<RecipeReadIdDto>> CreateRecipe(
         [FromServices] ICommandHandler<DeleteRecipeCommand> deleteRecipeCommandHandler )
     {
+        int? userId = GetUserIdFromAccessToken();
+        if ( userId is null )
```
это в аттрибут JwtAuthorization засунуть

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -62,20 +102,34 @@ public async Task<IActionResult> DeleteRecipe(
+            return Unauthorized();
+        }
+
+        UpdateRecipeCommand command = new() {
```
тоже автоматически надо маппить

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/JwtAuthorization/JwtAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace Recipes.Infrastructure.JwtAuthorization;
+
+public class JwtAuthorizationAttribute : Attribute, IAuthorizationFilter
```
это не должно быть в Infrastructure, это слой WebApi

VladDotMe replied:
Базовое правило - если у тебя в слое Infrastructure/Application/Domain есть какая-либо using  Microsoft.AspNetCore...  что-нибудь там - значит ты делаешь что-то неправильно)

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -11,37 +10,86 @@
-public class RecipesController : ControllerBase
+public class RecipesController( ITokenDecoder tokenDecoder ) : ControllerBase
 {
+    private int? GetUserIdFromAccessToken()
```
это в Extensions, на уровне WebApi

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -11,37 +10,86 @@
+            return null;
+        }
+
+        JwtSecurityToken token = tokenDecoder.DecodeToken( accessToken );
```
с токенДекодером будут проблемы, делай его статиком

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -11,37 +10,86 @@
     {
-        CreateRecipeCommand command = dto.Adapt<CreateRecipeCommand>();
+        int? userId = GetUserIdFromAccessToken();
+        if ( userId is null )
```
проверка бесполезная, у тебя на уровне аттрибута она уже есть.

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/JwtAuthorization/JwtAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,56 @@
+        string accessToken = context.HttpContext.Request.Headers[ "Access-Token" ];
+        if ( string.IsNullOrEmpty( accessToken ) )
+        {
+            context.Result = new ForbidResult();
```
Unathorized

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/JwtAuthorization/JwtAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,56 @@
+            return;
+        }
+
+        TokenSignatureVerificator tokenSignatureVerificator = new TokenSignatureVerificator();
```
тут тоже можно (нужно) получать классы через DI

VladDotMe commented:
Тоже не сразу заметил, Application/Infrastructure слой принимают IFormFile (зависимость от Microsoft.AspNet) - это неправильно. Нужно прокинуть абстракцию. Можно заюзать паттерн адаптер. Попробуй сделать, если не сможешь - можешь прийти ко мне или к чатужпт)

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/JwtAuthorization/JwtAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,56 @@
+        tokenSignatureVerificator.VerifySignature( accessToken, configuration.GetSecret() );
+        if ( !tokenSignatureVerificator.TokenIsValid )
+        {
+            context.Result = new ForbidResult();
```
UnauthorizedResult

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/JwtAuthorization/JwtAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,56 @@
+        DateTime expDate = DateTimeOffset.FromUnixTimeSeconds( token.Payload.Exp.Value ).UtcDateTime;
+        if ( DateTime.UtcNow > expDate )
+        {
+            context.Result = new ForbidResult();
```
UnauthorizedResult

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/JwtAuthorization/JwtAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,56 @@
+        Claim userIdClaim = token.Claims.FirstOrDefault( claim => claim.Type == "userId" );
+        if ( userIdClaim is null )
+        {
+            context.Result = new ForbidResult();
```
UnauthorizedResult

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/JwtAuthorization/JwtAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,56 @@
+            return;
+        }
+
+        if ( !int.TryParse( userIdClaim.Value, out int userId ) )
```
объединить с верхней проверкой и параметр discard _

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/JwtAuthorization/JwtAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        TokenDecoder tokenDecoder = new TokenDecoder();
+        JwtSecurityToken token = tokenDecoder.DecodeToken( accessToken );
```
DateTime expDate = DateTimeOffset.FromUnixTimeSeconds( token.Payload.Expiration.Value ).UtcDateTime;

Кстати, уверен, что token всегда != null? Если уверен, пропиши ! везде чтобы не ругался компилятор.  Либо проверь на null

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,186 @@
+[Route( "api/users" )]
+public class UsersController( ITokenDecoder tokenDecoder ) : ControllerBase
+{
+    private int? GetUserIdFromAccessToken()
```
удали

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,186 @@
+
+    [JwtAuthorization]
+    [HttpGet( "name" )]
+    public async Task<ActionResult<ReadUserDto>> GetUserNameById(
```
вроде же решили, что не используется и можно удалить?)

tanyginvv replied:
Переделал метод для получения имени пользователя при авторизации, чтобы отображать в headere на фронте.

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,186 @@
+        [FromServices] IQueryHandler<GetUserByIdQueryDto, GetUserByIdQuery> getUserByIdQueryHandler )
+    {
+        int? userId = GetUserIdFromAccessToken();
+        if ( userId is null )
```
проверка лишняя

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,186 @@
+            Login = user.Login,
+            Name = user.Name,
+            Description = user.Description,
+            RecipesCount = user.RecipesCount
```
RecipeCount правильно. С Count используется единственное число всегда.

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,186 @@
+        [FromServices] ICommandHandler<UpdateUserCommand> updateUserCommandHandler )
+    {
+        int? userId = GetUserIdFromAccessToken();
+        if ( userId is null )
```
тоже проверку убрать

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,186 @@
+        [FromBody] RegisterUserDto registerUserDto,
+        [FromServices] ICommandHandler<CreateUserCommand> createUserCommandHandler )
+    {
+        CreateUserCommand createUserCommand = new CreateUserCommand
```
mapster

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -11,37 +10,86 @@
+            return Unauthorized();
+        }
+
+        DeleteRecipeCommand command = new DeleteRecipeCommand
```
mapster

VladDotMe commented on backend/Recipes/Recipes.Application/UseCases/Recipes/Commands/CreateRecipe/CreateRecipeCommandHandler.cs:
```diff
@@ -26,10 +27,12 @@ public async Task<Result<RecipeIdDto>> HandleAsync( CreateRecipeCommand createRe
         Result validationResult = await validator.ValidateAsync( createRecipeCommand );
         if ( !validationResult.IsSuccess )
         {
+            imageTools.DeleteImage( createRecipeCommand.ImageUrl );
```
С этим тоже есть проблема, если ошибка произойдет ниже. Ну и в целом есть сейчас проблема с исключениями, которые могут возникнуть на уровне хендлеров при работе с базой, например. Сейчас в таком случае с бекенда выбросится 500я ошибка, что не очень хорошо. 

Варианта вижу два:
1. Middleware на обработку ошибок (скорее всего так все сделают)
2. Предпочительный как по мне - написать общий декоратор на все хендлеры, тогда впринципе не будет ситуация когда хендлер выбросит исключение.
Декоратор:
Во-первых будет валидировать (т.к. это общая лоигка во всех хендлерах и можно будет убрать ее из самих хенлдеров), Во-вторых отлавливать все ошибки try/catch и конвертировать их в Result., В-третьих, будет делать cleanup при необходимости (тут не уверен, что получитяс, возможно сильно проще в самом хендлере try/catch/finally и + проброс ошибки вверх). 
Ну и плюс ResultError лучше расширить и добавить кастомный тип, чтобы с бекенда возвращать разные статусы а не только BadRequest().  Можно заюзать библиотеку типа Scuptor, можно написать самому.
Если ничего не понял - спрашивай в телеге)

VladDotMe replied:
UPD: Я тут понял, что с декораторами я немного погорячился, куда лучше/проще сделать абстрактный класс + шаблонный метод.

tanyginvv committed:
Правки по комментам, пока без декоратора для хенделров

1) добавил mapster, где нужно было
2) вынес jwt авторизацию в слой апи
3) добавил адаптер для работы с картинками
4)также токендекодер сделал статиком с помощью адаптера( нельзя наследоваться статичному классу от интерфейса, либо я что-то не понял)
5) добавил проверку токена на null, также при авторизации теперь в контексте сохраняю userId из токена и использую его в контроллерах
6) вынес для метода получения рецептов в extensions для апи метод получения userId из токена( логика такая, что получить рецепты мы можем и без авторизации, но если в headerах встретим accessToken то получим оттуда id пользователя и покажем только его рецепты), не знаю на сколько эта логика корректна, но пока что так.

VladDotMe commented on backend/Recipes/Recipes.WebApi/JwtAuthorization/JwtAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,59 @@
+        {
+            context.Result = new UnauthorizedResult();
+            return;
+        }
```
if ( token?.Payload?.Expiration is null || DateTime.UtcNow > DateTimeOffset.FromUnixTimeSeconds( token.Payload.Expiration.Value ).UtcDateTime )

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -52,21 +70,26 @@ public async Task<IActionResult> DeleteRecipe(
     [HttpPut( "{id}" )]
-    public async Task<IActionResult> UpdateRecipe(
+    public async Task<ActionResult<Result>> UpdateRecipe(
         [FromRoute, Range( 1, int.MaxValue )] int id,
```
а зачем? у тебя в дтохе передается же айди

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -11,37 +10,56 @@
+            return Unauthorized();
+        }
+
+        DeleteRecipeCommand command = new DeleteRecipeCommand
```
все еще не мапстер..

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -11,37 +10,56 @@
-        [FromServices] IImageTools imageTool )
+        [FromServices] ICommandHandlerWithResult<CreateRecipeCommand, RecipeIdDto> createRecipeCommandHandler )
     {
+        if ( !HttpContext.Items.TryGetValue( "userId", out object userIdObj ) || userIdObj is not int userId )
```
убрать все return Unathorized() из методво контроллера, для получения юзерайди юзать Extensions. и использовать вида HttpContext.User.GetUserId();

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -95,10 +118,13 @@ public async Task<ActionResult<IEnumerable<RecipePartReadDto>>> GetRecipes(
     {
+        int userId = HttpContext.GetUserIdFromAccessToken( tokenDecoder ) ?? 0;
+
         GetRecipesQuery query = new GetRecipesQuery
```
тоже мапстер

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/RecipesController.cs:
```diff
@@ -95,10 +118,13 @@ public async Task<ActionResult<IEnumerable<RecipePartReadDto>>> GetRecipes(
         [FromQuery] int pageNumber = 1,
         [FromQuery] List<string> searchTerms = null )
     {
+        int userId = HttpContext.GetUserIdFromAccessToken( tokenDecoder ) ?? 0;
```
tokenDecoder не круто передавать, сделай его статическим просто

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Tokens/DecodeToken/TokenDecoderAdapter.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Recipes.Infrastructure.Tokens.DecodeToken;
+
+public class TokenDecoderAdapter : ITokenDecoder
```
зачем?? удалить этот класс, использовать напрямую статический метод в Extensions

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Tokens/DecodeToken/TokenDecoder.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Recipes.Infrastructure.Tokens.DecodeToken;
+
+public static class TokenDecoder
```
можно назвать TokenUtils

VladDotMe commented on backend/Recipes/Recipes.WebApi/JwtAuthorization/JwtAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,59 @@
+        }
+
+        JwtSecurityToken token;
+        try
```
давай лучше try/catch на весь метод повесим на всякий случай, мало ли чего упасть здесь может..  + логгирование добавить в случае падения ошибки. Используй дефолтный ILogger от майкрософта и прокинь через DI, потом можно будет либу накатить чтобы в файлы писать и т.д., но пока дефолтного хватит, который будет в консоль писать.

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/ImagesController.cs:
```diff
@@ -1,22 +1,27 @@
+        IFormFile image,
         [FromServices] IImageTools imageHelperTools )
     {
         if ( image is null )
```
есть аттрибут [NotNull]. Но проверь плз, что он выбрасывает ошикбу, потому что я не на 100% уверен

VladDotMe commented on backend/Recipes/Recipes.WebApi/Adapters/FormFileAdapter.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace Recipes.Infrastructure.ImageTools
+{
+    public class FormFileAdapter( IFormFile formFile ) : IFile
```
классно сделал) только поправь неймспейс

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/ImageTools/ImageHelper.cs:
```diff
@@ -1,90 +1,82 @@
+    public class ImageHelperTools : IImageTools
     {
-        if ( image is null )
+        private const string StorageUrl = "../Recipes.Infrastructure/Storage";
```
а еще, я бы путь законфигурил через appsettings.json и этот файл можно назвать FileImageTools (чтобы явно сказать, что мы сохраняем и достаем изображения из файлов, вдруг потом появится какой-нибудь DatabaseImageTools или CloudImageTools и т.д.)

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/ImageTools/ImageHelper.cs:
```diff
@@ -1,90 +1,82 @@
         {
-            return null;
-        }
+            if ( file is null || file.Length == 0 )
```
здесь не хватает try/catch тоже на весь метод, может упасть что-нибудь, все таки с I/O работаем, там всегда повышенный риск ошибок.

VladDotMe commented on backend/Recipes/Recipes.Application/Interfaces/IImageTools.cs:
```diff
@@ -1,10 +1,8 @@
 public interface IImageTools
 {
-    Task<string> SaveImageAsync( IFormFile image );
+    Task<string> SaveImageAsync( IFile image );
```
а еще так-то в можно тоже в Result все обернуть чтобы не проверять на null а проверять, на IsSuccess, но здесь в целом не так критично

VladDotMe commented on backend/Recipes/Recipes.WebApi/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,141 @@
+    public async Task<ActionResult<ReadUserDto>> GetUserNameById(
+        [FromServices] IQueryHandler<GetUserNameByIdQueryDto, GetUserNameByIdQuery> getUserNameByIdQueryHandler )
+    {
+        if ( !HttpContext.Items.TryGetValue( "userId", out object userIdObj ) || userIdObj is not int userId )
```
тут тоже везде не забудь переделать

VladDotMe commented on backend/Recipes/Recipes.WebApi/JwtAuthorization/JwtAuthorizationAttribute.cs:
```diff
@@ -0,0 +1,59 @@
+
+public class JwtAuthorizationAttribute : Attribute, IAuthorizationFilter
+{
+    private readonly ITokenDecoder tokenDecoder;
```
что-то здесь не так, поля не прокидываются

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/ConfigurationUtils/JwtOptions.cs:
пробежался еще по структуре, папку и неймспейс назови "OptionsS"

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/Entities/Recipes/RecipeConfiguration.cs:
А папку Entities лучше назвать DataAccess и в нее же можно рядом положить RecipesDbContext и UnitOfWork

VladDotMe commented:
А еще Validator из папки Validators перенеси в CQRSInterfaces, это же тоже к этому относится. Лишние папки плодить не очень хорошо.

tanyginvv committed:
Правки

tanyginvv committed:
правки по получению рецептов

tanyginvv committed:
вернул получение рецептов на GET

tanyginvv committed:
добавил удаление картинки рецепта, если что-то пошло неудачно при создании или его обновлении

tanyginvv committed:
вернул все как было ( переделывал для проверки работы удаления картинок)

tanyginvv committed:
мини-правка

VladDotMe commented:
Щас еще заметил, в Program.cs
string environmentName = Environment.GetEnvironmentVariable( "JSON_CONFIG_NAME" ) ?? "Development";

Здесь fallback на Development, хотя appsettings.Development.json нет, надо добавить. И лучше просто написать ?? "dev"

VladDotMe commented on backend/Recipes/Recipes.WebApi/Program.cs:
```diff
@@ -1,17 +1,36 @@
 
 WebApplicationBuilder builder = WebApplication.CreateBuilder( args );
 
+string environmentName = Environment.GetEnvironmentVariable( "JSON_CONFIG_NAME" ) ?? "Development";
```
Здесь fallback на Development, хотя appsettings.Development.json нет, надо добавить. И лучше просто назвать "dev" вместо "Development"

VladDotMe commented on backend/Recipes/Recipes.Application/CQRSInterfaces/ICommandHandler.cs:
```diff
@@ -9,4 +9,63 @@ public interface ICommandHandler<TCommand> where TCommand : class
+        Result validationResult = await validator.ValidateAsync( command );
+        if ( !validationResult.IsSuccess )
+        {
+            await HandleExceptionAsync( command );
```
Это не Exception, HandleException должен быть в нижнем catch блоке..

VladDotMe commented on backend/Recipes/Recipes.Application/CQRSInterfaces/ICommandHandler.cs:
```diff
@@ -9,4 +9,63 @@ public interface ICommandHandler<TCommand> where TCommand : class
+        Result validationResult = await validator.ValidateAsync( command );
+        if ( !validationResult.IsSuccess )
+        {
+            await HandleExceptionAsync( command );
```
И удалять изображение тебе нужно там же а не в случае валидации, т.к. может произойти какая-либо ошибка сохранения рецепта в базу и тогда тебе тоже нужно удалять изображение.

tanyginvv replied:
По сути у меня же сначала отправляется фото и я получаю урл, и если валидация не пройдена, я удаляю картинку, и если что-то пошло не так уже после пройденной валидации также ее удаляю. Таким образом, в двух местах нужно оставить, по-моему.

VladDotMe commented on backend/Recipes/Recipes.WebApi/Program.cs:
```diff
@@ -1,17 +1,36 @@
+{
+    options.AddPolicy( "AllowSpecificOrigin", policy =>
+    {
+        policy.WithOrigins( "http://localhost:3000" )
```
урл в конфиг

VladDotMe commented on backend/Recipes/Recipes.Application/CQRSInterfaces/ICommandHandler.cs:
```diff
@@ -9,4 +9,63 @@ public interface ICommandHandler<TCommand> where TCommand : class
+        Result validationResult = await validator.ValidateAsync( command );
+        if ( !validationResult.IsSuccess )
+        {
+            await HandleExceptionAsync( command );
```
Здесь аналогично предыдущему комменту

VladDotMe commented on backend/Recipes/Recipes.Application/CQRSInterfaces/ICommandHandler.cs:
```diff
@@ -9,4 +9,63 @@ public interface ICommandHandler<TCommand> where TCommand : class
     Task<Result<TResult>> HandleAsync( TCommand command );
+}
+
+public abstract class CommandBaseHandler<TCommand, TResult>( IAsyncValidator<TCommand> validator ) 
```
еще бы базовые классы вынести отдельно. В идеале 4 файла (все интерфейсы по разным файлам, все классы по разным файлам)

VladDotMe commented on backend/Recipes/Recipes.Application/CQRSInterfaces/ICommandHandler.cs:
```diff
@@ -9,4 +9,63 @@ public interface ICommandHandler<TCommand> where TCommand : class
+        }
+    }
+
+    protected abstract Task<Result<TResult>> HandleAsyncImpl( TCommand command );
```
Async в конец

tanyginvv replied:
Студия говорит, что можно использовать async только в методах, которые имеют тело, в командах async прописан

VladDotMe replied:
> Студия говорит, что можно использовать async только в методах, которые имеют тело, в командах async прописан

я про название метода

VladDotMe commented:
У BaseRepository добавь неймспейс

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/ImageTools/FileToolConfiguration.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace Recipes.Infrastructure.ImageTools;
+
+public class FileToolConfiguration( IOptions<FileToolsOptions> toolOptions) : IFileToolConfiguration
```
лишний класс же, зачем? IOptions сам по себе мокается, если что)

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/ImageTools/FileToolsOptions.cs:
```diff
@@ -0,0 +1,6 @@
+
+public class FileToolsOptions
+{
+    public string StorageUrl { get; set; }
```
Ну и это можно в папку Options перенести рядом со всеми

VladDotMe commented:
IFilter лучше положить в папку Filters

VladDotMe commented on backend/Recipes/Recipes.Application/Interfaces/IFileToolConfiguration.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace Recipes.Application.Interfaces;
+
+public interface IFileToolConfiguration
```
этот удалить

tanyginvv committed:
Правки по комментам

tanyginvv committed:
добавил первый тест кейс

tanyginvv committed:
правки

tanyginvv committed:
правки с конфигурацией токена

VladDotMe commented on backend/Recipes/Recipes.Infrastructure/ImageTools/FileImageTools.cs:
```diff
@@ -0,0 +1,98 @@
+
+namespace Recipes.Infrastructure.ImageTools;
+
+public class FileImageTools( IOptions<FileToolsOptions> fileTool ) : IImageTools
```
fileToolsOptions

VladDotMe commented on backend/Recipes/Recipes.WebApi/appsettings.prod.json:
```diff
@@ -1,9 +1,14 @@
+        "Recipes": "Server=(localDB)\\MSSQLLocalDB;Database=Recipes;Trusted_Connection=true;TrustServerCertificate=True;"
+    },
+    "JwtOptions": {
+        "Secret": "96b14af9-337d-44e0-ae81-f6af75ba9b38",
```
кстати, на разных средах обычно разные токены. Ну это так, на будущее, пока у тебя тут все равно везде тестовые данные.

tanyginvv committed:
last правки
